File: js/setup/GamePadInput.js
// js/setup/gamepadInput.js
// pollGamepad(): reads the first connected gamepad's current state and
// translates it into the same shared state the rest of the game
// already reads — state.keys['ArrowLeft']/['ArrowRight'] for walking
// (exactly what Player.move() already checks for keyboard input, so
// that function needed zero changes), a fire-held flag gameLoop checks
// alongside state.mouseDown, state.gamepadAimActive/X/Y for
// Player.js's computeLeftArmAimAngle to read as an aim source,
// state.gamepadIsActiveDevice so that same method knows to default to
// aiming straight ahead (rather than falling back to the mouse) when
// the stick is centered but the player is still clearly using the
// controller, R1/L1/Circle button presses that call straight into
// Player.js's tryLockTargetNext()/tryLockTargetPrevious()/
// clearLockTarget() for the hard-lock system, and Triangle toggling
// state.vatsActive for V.A.T.S. (see game.js's timeScale easing and
// effects/VatsOverlay.js for the rest of that system).
//
// Unlike keyboard/mouse (event-driven — inputHandlers.js just attaches
// listeners once), the Gamepad API has no press/release/move events at
// all for stick or button state — the only way to know the CURRENT
// state is to poll navigator.getGamepads() fresh every single frame.
// That's why this is a function called every frame from gameLoop() in
// game.js, rather than a one-time attachInputHandlers()-style setup
// call the way every other input source in this game works.
import { state } from '../state.js';
import { tryRestartOrAdvance } from './inputHandlers.js';

// Standard gamepad mapping button indices (https://w3c.github.io/gamepad/#remapping).
// IMPORTANT GOTCHA: the spec's button NAMES use Xbox terminology — its
// button index 2 is literally labeled "X" there, but that's Xbox's own
// X button, which sits in the LEFT face-button position (PlayStation's
// Square in that same position). PS5's actual Cross/X button sits in
// the BOTTOM face-button position, which is index 0 in the standard
// mapping regardless of controller brand. Getting this backwards is a
// very easy, very common mistake.
const BUTTON_X_CROSS = 0;
const BUTTON_CIRCLE = 1;
const BUTTON_SQUARE = 2;
const BUTTON_TRIANGLE = 3;
const BUTTON_R2 = 7;
const BUTTON_L2 = 6;
const BUTTON_R1 = 5;
const BUTTON_L1 = 4;
const BUTTON_DPAD_UP = 12;
const BUTTON_DPAD_DOWN = 13;

const STICK_DEADZONE = 0.2; // ignore small stick drift near center, on both sticks

// Tracked across frames so the X button's jump can be edge-triggered
// (fires once per press, not once per frame it's held) — gamepad
// buttons are polled, not events, so detecting "just pressed this
// frame" (this frame true, last frame false) has to be done by hand.
let lastXPressed = false;
let lastL2Pressed = false;
let lastR1Pressed = false;
let lastL1Pressed = false;
let lastCirclePressed = false;
let lastTrianglePressed = false;

// Tracked so the left stick can CLEAR state.keys['ArrowLeft']/
// ['ArrowRight'] when released, WITHOUT ever clearing a flag the
// keyboard itself set — only clearing a flag this module itself most
// recently set to true. Without this, an idle/centered stick would
// unconditionally force both flags false every single frame, which
// would silently break keyboard movement entirely the moment any
// gamepad was connected, even if the player never touched it.
let gamepadHeldLeft = false;
let gamepadHeldRight = false;
let gamepadHeldZoomIn = false;
let gamepadHeldZoomOut = false;

export function pollGamepad() {
  const gamepads = navigator.getGamepads();
  // Just the first connected gamepad — this game has no multiplayer or
  // gamepad-selection concept, so "whichever one is first" is a
  // reasonable, simple default.
  const gp = Array.from(gamepads).find(g => g);
  if (!gp) {
    state.gamepadAimActive = false;
    state.gamepadFireHeld = false;
    state.gamepadRunHeld = false;
    state.gamepadIsActiveDevice = false;
    return;
  }

  // On the GAME OVER / level-complete screens, X is the ONLY thing
  // this function should do — walk/fire/pull all need to be skipped
  // entirely here, not just left harmless, since L2 below calls
  // trySelectPullTarget() directly (immediately, not via a flag
  // checked later), which would otherwise still be able to launch the
  // player and set a pull target while these screens are up. Mirrors
  // how the mouse path already behaves: inputHandlers.js's mousedown
  // handler checks tryRestartOrAdvance() FIRST and returns immediately
  // if it did something, before ever reaching the right-click pull
  // logic.
  if (state.gameOver || state.levelComplete) {
    const xPressed = gp.buttons[BUTTON_X_CROSS]?.pressed || false;
    if (xPressed && !lastXPressed) {
      tryRestartOrAdvance();
    }
    lastXPressed = xPressed;
    return;
  }

  // --- Left stick: walk left/right ---
  const leftX = gp.axes[0] || 0;
  if (leftX < -STICK_DEADZONE) {
    state.keys['ArrowLeft'] = true;
    gamepadHeldLeft = true;
    if (gamepadHeldRight) { state.keys['ArrowRight'] = false; gamepadHeldRight = false; }
  } else if (leftX > STICK_DEADZONE) {
    state.keys['ArrowRight'] = true;
    gamepadHeldRight = true;
    if (gamepadHeldLeft) { state.keys['ArrowLeft'] = false; gamepadHeldLeft = false; }
  } else {
    if (gamepadHeldLeft) { state.keys['ArrowLeft'] = false; gamepadHeldLeft = false; }
    if (gamepadHeldRight) { state.keys['ArrowRight'] = false; gamepadHeldRight = false; }
  }

  // --- D-pad up/down: zoom in/out ---
  // Feeds into the SAME state.keys['+']/['-'] flags the keyboard's own
  // zoom keys already set — game.js's gameLoop reads those directly
  // (held-continuous, same ZOOM_STEP_PER_FRAME either way), so this
  // needed zero changes there, same reasoning as the left stick above.
  // Same "only clear a flag this module itself set" bookkeeping too,
  // so an idle D-pad can't stomp on an actual keyboard +/- press.
  const dpadUpPressed = gp.buttons[BUTTON_DPAD_UP]?.pressed || false;
  const dpadDownPressed = gp.buttons[BUTTON_DPAD_DOWN]?.pressed || false;
  if (dpadUpPressed) {
    state.keys['+'] = true;
    gamepadHeldZoomIn = true;
  } else if (gamepadHeldZoomIn) {
    state.keys['+'] = false;
    gamepadHeldZoomIn = false;
  }
  if (dpadDownPressed) {
    state.keys['-'] = true;
    gamepadHeldZoomOut = true;
  } else if (gamepadHeldZoomOut) {
    state.keys['-'] = false;
    gamepadHeldZoomOut = false;
  }

  // --- X (Cross): jump / ground pound ---
  // Mirrors the spacebar's own keydown behavior in inputHandlers.js.
  // (GAME OVER / level-complete restart handling for X lives in the
  // dedicated early branch above instead, not here.)
  const xPressed = gp.buttons[BUTTON_X_CROSS]?.pressed || false;
  if (xPressed && !lastXPressed && state.player && state.player.mode !== "maze") {
    if (state.player.onSurface) state.player.jump();
    else state.player.tryGroundPound();
  }
  lastXPressed = xPressed;

  // --- R2: fire ---
  // A continuous-hold flag, same pattern as state.mouseDown — gameLoop
  // checks this every frame and calls shootFireball(), which already
  // has its own internal cooldown (this.fireCooldown), so holding this
  // down is safe and doesn't need its own rate-limiting here.
  state.gamepadFireHeld = gp.buttons[BUTTON_R2]?.pressed || false;

  // --- Triangle: toggle V.A.T.S. ---
  // Edge-triggered — a press turns it on if off, or off if on (a
  // second Triangle press is the manual "back out without firing"
  // escape hatch; the automatic exits — firing or a successful pull —
  // live inside shootFireball()/trySelectPullTarget() themselves in
  // Player.js, since gamepadInput.js only INITIATES those calls each
  // frame and has no way to know whether they actually succeeded).
  // Only toggleable in "space" mode — locking onto world objects
  // doesn't mean anything inside a maze/platform interior — and forced
  // back off if the player somehow leaves space mode while it's
  // active, so the slowdown/overlay can never persist somewhere it
  // doesn't make sense.
  const trianglePressed = gp.buttons[BUTTON_TRIANGLE]?.pressed || false;
  if (state.player && state.player.mode === "space") {
    if (trianglePressed && !lastTrianglePressed) {
      state.vatsActive = !state.vatsActive;
    }
  } else if (state.vatsActive) {
    state.vatsActive = false;
  }
  lastTrianglePressed = trianglePressed;

  // --- Square: run ---
  // A continuous-hold flag, same pattern as gamepadFireHeld above —
  // read directly by Player.move() to boost walking speed while held,
  // and by spawnWalkDust() to kick up more dust per footstep while
  // running.
  state.gamepadRunHeld = gp.buttons[BUTTON_SQUARE]?.pressed || false;

  // --- R1: hard lock, next (clockwise) ---
  // --- L1: hard lock, previous (counter-clockwise) ---
  // Both edge-triggered — each press cycles to a candidate (see
  // Player.js's tryLockTargetNext/tryLockTargetPrevious and
  // TargetLock.js's selectNextLockTarget/selectPreviousLockTarget),
  // neither is a hold-to-lock button. L1 is the mirror of R1 — in the
  // common case, pressing one then the other undoes the first press.
  const r1Pressed = gp.buttons[BUTTON_R1]?.pressed || false;
  if (r1Pressed && !lastR1Pressed && state.player) {
    state.player.tryLockTargetNext();
  }
  lastR1Pressed = r1Pressed;

  const l1Pressed = gp.buttons[BUTTON_L1]?.pressed || false;
  if (l1Pressed && !lastL1Pressed && state.player) {
    state.player.tryLockTargetPrevious();
  }
  lastL1Pressed = l1Pressed;

  // --- Circle: release lock, and end V.A.T.S. ---
  // A natural dual-purpose "cancel" press — releasing whatever's
  // locked and backing out of V.A.T.S. (if active) are both "never
  // mind" actions, so one button doing both reads as consistent rather
  // than needing a second, separate cancel button.
  const circlePressed = gp.buttons[BUTTON_CIRCLE]?.pressed || false;
  if (circlePressed && !lastCirclePressed && state.player) {
    state.player.clearLockTarget();
    state.vatsActive = false;
  }
  lastCirclePressed = circlePressed;

  // --- L2: pull beam ---
  // Mirrors the mouse's right-click UX exactly: press to initiate a
  // pull (edge-triggered — trySelectPullTarget() does a one-time
  // launch-off-the-surface kick, so this must fire once per press, not
  // every frame it's held), release to let go. Prefers the current
  // hard lock if one's active (this is in fact the main reason the
  // lock system exists — locking on specifically to make pulling more
  // reliable/deliberate than the raycast alone), falling back to
  // whatever the raycast most recently found (state.player.
  // aimTargetObject — the same object the red X would be resting on,
  // were it not suppressed while locked) otherwise. Passed as
  // trySelectPullTarget's explicit-target parameter so this reuses the
  // exact same validation/launch logic the mouse path already has.
  const l2Pressed = gp.buttons[BUTTON_L2]?.pressed || false;
  if (state.player) {
    if (l2Pressed && !lastL2Pressed) {
      state.player.trySelectPullTarget(state.player.lockedTarget || state.player.aimTargetObject);
    } else if (!l2Pressed && lastL2Pressed) {
      state.player.clearPullTarget();
    }
  }
  lastL2Pressed = l2Pressed;

  // --- Right stick: aim ---
  // Stored as a normalized DIRECTION (not a position — sticks don't
  // have one), read by Player.js's computeLeftArmAimAngle as a second-
  // priority aim source (after pullTarget, before the mouse fallback).
  // gamepadAimActive naturally goes true/false as the stick is pushed/
  // released, which is also what Player.js's mouseIdle computation
  // reads to know whether gamepad aim should override the mouse-idle
  // timeout.
  const rightX = gp.axes[2] || 0;
  const rightY = gp.axes[3] || 0;
  const rightMag = Math.hypot(rightX, rightY);
  if (rightMag > STICK_DEADZONE) {
    state.gamepadAimActive = true;
    state.gamepadAimX = rightX / rightMag;
    state.gamepadAimY = rightY / rightMag;
  } else {
    state.gamepadAimActive = false;
  }

  // --- Tracks whether the gamepad is the player's CURRENTLY preferred
  // input device (used more recently than the mouse) ---
  // Broader than gamepadAimActive (which only reflects "is the right
  // stick specifically pushed THIS exact frame") — this stays true
  // across frames where the stick is momentarily centered but the
  // player is still clearly playing with the controller (e.g. walking
  // with the left stick alone). Read by Player.js's
  // computeLeftArmAimAngle to decide whether "stick is centered" should
  // fall back to aiming straight ahead (gamepad play) or to the mouse
  // position (mouse play) — see that method's own comment for why this
  // distinction matters.
  const anyButtonPressed = gp.buttons.some(b => b.pressed);
  const anyStickActive = Math.abs(leftX) > STICK_DEADZONE || rightMag > STICK_DEADZONE;
  if (anyButtonPressed || anyStickActive) {
    state.lastGamepadInputTime = Date.now();
  }
  state.gamepadIsActiveDevice = (state.lastGamepadInputTime || 0) > (state.lastMouseMoveTime || 0);
}

File: lua/setup/GamePadInput.lua
-- lua/setup/gamepadInput.lua
--
-- Direct translation of js/setup/gamepadInput.js. Called from
-- love.update(dt), once per frame — same polling architecture as the
-- original, and for the same reason the original's own comment gives:
-- sticks have no "moved" event in either the browser Gamepad API or
-- LÖVE's own love.joystick module, so the only way to know CURRENT
-- stick position is to ask fresh every frame. (LÖVE does have
-- love.gamepadpressed/love.gamepadreleased event callbacks, unlike
-- the raw browser Gamepad API — but since sticks force polling
-- regardless, and the original's whole edge-detection design
-- [lastXPressed etc.] is already built around comparing this-frame
-- vs last-frame state, polling everything uniformly here is the more
-- faithful port, not a workaround.)
--
-- Real API differences worth knowing, all confirmed before writing
-- this rather than assumed:
--
--   Triggers (L2/R2) are ANALOG AXES in LÖVE ("triggerleft"/
--   "triggerright", read via getGamepadAxis, values roughly 0 to 1),
--   NOT buttons with a .pressed boolean the way the browser Gamepad
--   API exposes them. TRIGGER_PRESS_THRESHOLD below is what turns
--   that analog value into the same press/release boolean the
--   original's R2 (fire) and L2 (pull) logic needs — tune this
--   constant against your actual DualSense if the trigger feels like
--   it engages too early or too late; I can't test the physical feel
--   myself.
--
--   Button NAMES, not indices — LÖVE's a/b/x/y are named after their
--   XBOX POSITIONS (LÖVE's own docs: "correspond as closely as
--   possible to the layout of a standard Xbox 360 controller"), the
--   same Xbox-position-not-label convention the original's own BUTTON_*
--   comment already warns about for the browser API's numeric indices.
--   Mapped here by POSITION, matching the original's own reasoning
--   exactly: a=bottom=Cross, b=right=Circle, x=left=Square, y=top=
--   Triangle, leftshoulder=L1, rightshoulder=R1, dpup/dpdown=D-pad.
--
--   DualSense support: SDL (which LÖVE is built on) added explicit
--   DualSense recognition in version 2.0.14 (Dec 2020) — confirmed via
--   SDL's own release notes, not assumed — and LÖVE 11.x bundles SDL
--   well past that point, so the buttons/sticks/triggers this file
--   actually needs should work correctly out of the box. Peripheral
--   features (LED color, adaptive trigger rumble, the touchpad, the
--   mic button) have had a rockier history across SDL versions, but
--   none of those are used here.
--
--   Only the FIRST connected, gamepad-recognized joystick is used —
--   same "no multiplayer/selection concept" reasoning as the
--   original. love.joystick.getJoysticks() doesn't leave null gaps
--   for disconnected slots the way the browser's sparse gamepads
--   array can, so this is a plain array scan, not a `.find(g => g)`.
--
--   Date.now() (milliseconds) becomes love.timer.getTime() (seconds)
--   here too, same as inputHandlers.lua — and specifically the SAME
--   clock, so state.lastGamepadInputTime (set here) and
--   state.lastMouseMoveTime (set in inputHandlers.lua) stay directly
--   comparable, which state.gamepadIsActiveDevice's own comparison
--   below depends on.

local state = require("lua.state")
local inputHandlers = require("lua.setup.inputHandlers")
local tryRestartOrAdvance = inputHandlers.tryRestartOrAdvance

-- Positional mapping — see this file's own header comment for why
-- these specific names correspond to these specific PS buttons.
local BUTTON_X_CROSS = "a"
local BUTTON_CIRCLE = "b"
local BUTTON_SQUARE = "x"
local BUTTON_TRIANGLE = "y"
local BUTTON_R1 = "rightshoulder"
local BUTTON_L1 = "leftshoulder"
local BUTTON_DPAD_UP = "dpup"
local BUTTON_DPAD_DOWN = "dpdown"

local ALL_FACE_AND_SHOULDER_BUTTONS = {
  BUTTON_X_CROSS, BUTTON_CIRCLE, BUTTON_SQUARE, BUTTON_TRIANGLE,
  BUTTON_R1, BUTTON_L1, BUTTON_DPAD_UP, BUTTON_DPAD_DOWN, "dpleft", "dpright",
}

local STICK_DEADZONE = 0.2      -- ignore small stick drift near center, on both sticks
local TRIGGER_PRESS_THRESHOLD = 0.5  -- analog trigger value counted as "pressed" — tune against real hardware feel

-- Tracked across frames so buttons can be edge-triggered (fires once
-- per press, not once per frame held) — same reasoning and same
-- variable-per-button structure as the original.
local lastXPressed = false
local lastL2Pressed = false
local lastR1Pressed = false
local lastL1Pressed = false
local lastCirclePressed = false
local lastTrianglePressed = false

-- Tracked so the stick/d-pad can clear state.keys[...] on release
-- WITHOUT ever clearing a flag the keyboard itself set — same
-- bookkeeping and same reasoning as the original.
local gamepadHeldLeft = false
local gamepadHeldRight = false
local gamepadHeldZoomIn = false
local gamepadHeldZoomOut = false

local function maybeAutoEnterVats()
  if state.vatsAutoEnterOnLock and not state.vatsActive and state.player and state.player.mode == "space" then
    state.vatsActive = true
  end
end

local function pollGamepad(dt)
  local joysticks = love.joystick.getJoysticks()
  local gp = nil
  for _, joystick in ipairs(joysticks) do
    if joystick:isGamepad() then
      gp = joystick
      break
    end
  end

  if not gp then
    state.gamepadAimActive = false
    state.gamepadFireHeld = false
    state.gamepadRunHeld = false
    state.gamepadIsActiveDevice = false
    return
  end

  -- On the GAME OVER / level-complete screens, X is the ONLY thing
  -- this function should do — same early-return, same reasoning as
  -- the original.
  if state.gameOver or state.levelComplete then
    local xPressed = gp:isGamepadDown(BUTTON_X_CROSS)
    if xPressed and not lastXPressed then
      tryRestartOrAdvance()
    end
    lastXPressed = xPressed
    return
  end

  -- --- Left stick: walk left/right ---
  local leftX = gp:getGamepadAxis("leftx") or 0
  if leftX < -STICK_DEADZONE then
    state.keys['ArrowLeft'] = true
    gamepadHeldLeft = true
    if gamepadHeldRight then state.keys['ArrowRight'] = false; gamepadHeldRight = false end
  elseif leftX > STICK_DEADZONE then
    state.keys['ArrowRight'] = true
    gamepadHeldRight = true
    if gamepadHeldLeft then state.keys['ArrowLeft'] = false; gamepadHeldLeft = false end
  else
    if gamepadHeldLeft then state.keys['ArrowLeft'] = false; gamepadHeldLeft = false end
    if gamepadHeldRight then state.keys['ArrowRight'] = false; gamepadHeldRight = false end
  end

  -- --- D-pad up/down: zoom in/out ---
  local dpadUpPressed = gp:isGamepadDown(BUTTON_DPAD_UP)
  local dpadDownPressed = gp:isGamepadDown(BUTTON_DPAD_DOWN)
  if dpadUpPressed then
    state.keys['+'] = true
    gamepadHeldZoomIn = true
  elseif gamepadHeldZoomIn then
    state.keys['+'] = false
    gamepadHeldZoomIn = false
  end
  if dpadDownPressed then
    state.keys['-'] = true
    gamepadHeldZoomOut = true
  elseif gamepadHeldZoomOut then
    state.keys['-'] = false
    gamepadHeldZoomOut = false
  end

  -- --- X (Cross): jump / wall jump / ground pound ---
  -- Player:jump() itself branches on grounded-launch vs. wall-jump kick
  -- (see Player.lua) — routed here together the same way InputHandlers.lua
  -- routes the keyboard Space key, since both are "this button performs a
  -- jump," as opposed to the ground-pound fallback for mid-air with
  -- nothing to jump off of.
  local xPressed = gp:isGamepadDown(BUTTON_X_CROSS)
  if xPressed and not lastXPressed and state.player and state.player.mode ~= "maze" and not state.introLocked then
    if state.player.onSurface or state.player.touchingWall then state.player:jump()
    else state.player:tryGroundPound() end
  end
  lastXPressed = xPressed

  -- --- R2: fire ---
  -- Analog trigger, thresholded — see this file's own header comment.
  state.gamepadFireHeld = (gp:getGamepadAxis("triggerright") or 0) > TRIGGER_PRESS_THRESHOLD

  -- --- Triangle: toggle V.A.T.S. ---
  local trianglePressed = gp:isGamepadDown(BUTTON_TRIANGLE)
  if state.player and state.player.mode == "space" then
    if trianglePressed and not lastTrianglePressed then
      state.vatsActive = not state.vatsActive
    end
  elseif state.vatsActive then
    state.vatsActive = false
  end
  lastTrianglePressed = trianglePressed

  -- --- Square: run ---
  state.gamepadRunHeld = gp:isGamepadDown(BUTTON_SQUARE)

  -- --- R1: hard lock, next --- / --- L1: hard lock, previous ---
  local r1Pressed = gp:isGamepadDown(BUTTON_R1)
  if r1Pressed and not lastR1Pressed and state.player then
    maybeAutoEnterVats()
    state.player:tryLockTargetNext()
  end
  lastR1Pressed = r1Pressed

  local l1Pressed = gp:isGamepadDown(BUTTON_L1)
  if l1Pressed and not lastL1Pressed and state.player then
    maybeAutoEnterVats()
    state.player:tryLockTargetPrevious()
  end
  lastL1Pressed = l1Pressed

  -- --- Circle: release lock, and end V.A.T.S. ---
  local circlePressed = gp:isGamepadDown(BUTTON_CIRCLE)
  if circlePressed and not lastCirclePressed and state.player then
    state.player:clearLockTarget()
    state.vatsActive = false
  end
  lastCirclePressed = circlePressed

  -- --- L2: pull beam ---
  -- Analog trigger, thresholded — same reasoning as R2 above.
  local l2Pressed = (gp:getGamepadAxis("triggerleft") or 0) > TRIGGER_PRESS_THRESHOLD
  if state.player then
    if l2Pressed and not lastL2Pressed then
      state.player:trySelectPullTarget(state.player.lockedTarget or state.player.aimTargetObject)
    elseif not l2Pressed and lastL2Pressed then
      state.player:clearPullTarget()
    end
  end
  lastL2Pressed = l2Pressed

  -- --- Right stick: aim (or, while in V.A.T.S., drives VatsCursor.lua's
  -- own on-screen cursor instead — see that file) ---
  local rightX = gp:getGamepadAxis("rightx") or 0
  local rightY = gp:getGamepadAxis("righty") or 0
  local rightMag = math.sqrt(rightX * rightX + rightY * rightY)
  if rightMag > STICK_DEADZONE then
    state.gamepadAimActive = true
    state.gamepadAimX = rightX / rightMag
    state.gamepadAimY = rightY / rightMag
    -- Raw (NOT unit-normalized) deflection, deadzone applied but actual
    -- push distance preserved — gamepadAimX/Y above throw that away on
    -- purpose (a fixed-length aim ray only cares about direction), but
    -- VatsCursor.lua's cursor wants proportional speed: a light nudge
    -- should crawl, a full push should move at full speed.
    state.gamepadRightStickX = rightX
    state.gamepadRightStickY = rightY
  else
    state.gamepadAimActive = false
    state.gamepadRightStickX = 0
    state.gamepadRightStickY = 0
  end

  -- --- Tracks whether the gamepad is the player's CURRENTLY
  -- preferred input device ---
  local anyButtonPressed = gp:isGamepadDown(ALL_FACE_AND_SHOULDER_BUTTONS)
    or (gp:getGamepadAxis("triggerleft") or 0) > TRIGGER_PRESS_THRESHOLD
    or (gp:getGamepadAxis("triggerright") or 0) > TRIGGER_PRESS_THRESHOLD
  local anyStickActive = math.abs(leftX) > STICK_DEADZONE or rightMag > STICK_DEADZONE
  if anyButtonPressed or anyStickActive then
    state.lastGamepadInputTime = love.timer.getTime()
  end
  state.gamepadIsActiveDevice = (state.lastGamepadInputTime or 0) > (state.lastMouseMoveTime or 0)
end

return {
  pollGamepad = pollGamepad
}